vowpalwabbit/features_data.h

```diff
@@ -0,0 +1,147 @@
+#include <array>
+#include <memory>
+#include <type_traits>
+
+/*
+ * class features_data
+ * Description:
+ *   This data structure manages access and lifetime of the features data used in various VW reductions.
+ *     New types must be created with the templatized create() function call, attempting to access
+ *     uninitialized data will return a nullptr.
+ *
+ *   All contained data types MUST instantiate a `clear()` function that readies the object for reuse.
+ *
+ *   Calling create() on an existing element will clear and overwrite the existing data
+ *
+ * Member functions:
+ *   T* create(constructor_parameters);
+ *   T* get();
+ *   void clear();
+ *
+ * Data structure usage:
+ *   Add new data structure to the available types:
+ *     Add a new line to the TYPE_MAPPINGS macro
+ *
+ *   Create an instance of a data structure:
+ *     features_data fd;
+ *     fd.create<data_type>(constructor_parameters);
+ *
+ *   Accessing an instance of existing data:
+ *     auto data = fd.get<data_type>();
+ */
+
+/////////////////////////////////////////////
+// sample placeholder structs. For demonstrative purposes only
+struct basic_features{
+basic_features() : x(0) {}
+  int x;
+  void clear();
+};
+struct ccb_features
+{
+  int x;
+  float y;
+  double z;
+  void clear();
+};
+/////////////////////////////////////////////
+
+/*
+ * Escape the last line and add new data structure->enum mappings to the end of this list
+ */
+#define TYPE_MAPPINGS        \
+  MAP(basic_features, basic) \
+  MAP(ccb_features, ccb)
```

Comment:
Contributors will need to escape the last line before adding a new one (otherwise you get compiler warnings). Failing to do so will result in (somewhat cryptic) compiler errors.

Is that acceptable?